addOns/authhelper/src/main/java/org/zaproxy/addon/authhelper/client/ClientScriptBasedAuthenticationMethodType.java

```diff
@@ -0,0 +1,750 @@
+import java.util.Set;
+import javax.swing.DefaultComboBoxModel;
+import javax.swing.JButton;
+import javax.swing.JLabel;
+import javax.swing.JOptionPane;
+import javax.swing.JPanel;
+import org.apache.commons.configuration.Configuration;
+import org.apache.commons.configuration.ConfigurationException;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jdesktop.swingx.JXComboBox;
+import org.jdesktop.swingx.decorator.FontHighlighter;
+import org.jdesktop.swingx.renderer.DefaultListRenderer;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.control.Control;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.network.HttpRequestHeader;
+import org.parosproxy.paros.view.View;
+import org.zaproxy.addon.authhelper.AuthUtils;
+import org.zaproxy.addon.authhelper.SessionManagementRequestDetails;
+import org.zaproxy.addon.authhelper.SessionToken;
+import org.zaproxy.addon.network.server.HttpMessageHandler;
+import org.zaproxy.zap.authentication.AbstractAuthenticationMethodOptionsPanel;
+import org.zaproxy.zap.authentication.AuthenticationCredentials;
+import org.zaproxy.zap.authentication.AuthenticationHelper;
+import org.zaproxy.zap.authentication.AuthenticationIndicatorsPanel;
+import org.zaproxy.zap.authentication.AuthenticationMethod;
+import org.zaproxy.zap.authentication.AuthenticationMethodType;
+import org.zaproxy.zap.authentication.GenericAuthenticationCredentials;
+import org.zaproxy.zap.authentication.ScriptBasedAuthenticationMethodType;
+import org.zaproxy.zap.extension.script.ExtensionScript;
+import org.zaproxy.zap.extension.script.ScriptWrapper;
+import org.zaproxy.zap.extension.zest.ZestAuthenticationRunner;
+import org.zaproxy.zap.model.Context;
+import org.zaproxy.zap.session.SessionManagementMethod;
+import org.zaproxy.zap.session.WebSession;
+import org.zaproxy.zap.users.User;
+import org.zaproxy.zap.utils.EncodingUtils;
+import org.zaproxy.zap.utils.ZapHtmlLabel;
+import org.zaproxy.zap.view.DynamicFieldsPanel;
+import org.zaproxy.zap.view.LayoutHelper;
+import org.zaproxy.zest.core.v1.ZestScript;
+
+public class ClientScriptBasedAuthenticationMethodType extends ScriptBasedAuthenticationMethodType {
+
+    public static final int METHOD_IDENTIFIER = 8;
+
+    private static final Logger LOGGER =
+            LogManager.getLogger(ClientScriptBasedAuthenticationMethodType.class);
+
+    private ExtensionScript extensionScript;
+
+    private HttpMessageHandler handler;
+    private HttpMessage authMsg;
+    private HttpMessage fallbackMsg;
+    private int firstHrefId;
+
+    public ClientScriptBasedAuthenticationMethodType() {}
+
+    private HttpMessageHandler getHandler(Context context) {
+        if (handler == null) {
+            handler =
+                    (ctx, msg) -> {
```

Comment:
This is duplicating `BrowserBasedAuthenticationMethodType`'s handler, better not.

Reply:
Ok, shall I create an "internal" package and generic ClientSideHandler class?

Reply:
Definitely under an internal package, that's not be used by other add-ons.